### PR TITLE
docs: add additional RBD image features to docs and YAML files

### PR DESCRIPTION
**Pull request by anthonyeleven**
Various files are outdated in that they limit RBD features to only `layering`.  `fast-diff` and `object-map` are beneficial on the Ceph back end for accelerating commands including `rbd du`.  Also add `exclusive-lock` and `deep-flatten` to fill out the conventional set of features.  Added note about this set being equivalent to the OR'd bitfield value 63, and that KRBD in the Linux kernel supports these additional features as of 5.4.

Signed-off-by: Anthony D'Atri <anthonyeleven@users.noreply.github.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: deploy/examples/import-external-cluster.sh
Comment (Rakshith-R) [inline deploy/examples/import-external-cluster.sh:199]: How about adding a env var to pass these extra imageFeatures. 
We don't need to add them always since they require kernel support ?

**Review by Rakshith-R — CHANGES_REQUESTED** [review]

File: deploy/examples/import-external-cluster.sh
Comment (anthonyeleven) [reply deploy/examples/import-external-cluster.sh:199]: That was a question I had -- does CSI use KRBD mounts, or librbd?

Comment (travisn) [reply deploy/examples/import-external-cluster.sh:199]: krbd is used by default. To use librbd you could set [mounter: rbd-nbd](https://github.com/rook/rook/blob/master/deploy/examples/csi/rbd/storageclass.yaml#L78-L81) in the storage class. However, it's not recommended for production as explained in that link.

Comment (anthonyeleven) [reply deploy/examples/import-external-cluster.sh]: Thanks.  I did find some references to librbd in the source but wasn't sure what actually got used in production.

My reasoning is that most people deploying `master` or the forthcoming 1.10 will be using a 5.4/equivalent or later kernel, though I may be mistaken.  5.4 came out at the end of 2019.  That said, who knows what's in any given RHEL/CentOS kernel?  And some people are stuck on older kernels for other reasons.

I've refactored this script, PTAL.

File: Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md
Comment (travisn) [inline Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md:66]: nit: blank line

File: Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md
Comment (travisn) [inline Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md:62]: Is there a need to know the numeric value?
```suggestion
    # RBD image features
```

File: deploy/charts/rook-ceph-cluster/values.yaml
Comment (travisn) [inline deploy/charts/rook-ceph-cluster/values.yaml:427]: Any reason not to make this consistent with the example storageclass.yaml?

File: deploy/examples/import-external-cluster.sh
Comment (travisn) [inline deploy/examples/import-external-cluster.sh:39]: copy-paste error? :) 
```suggestion
  if [ -z "$ROOK_RBD_FEATURES" ]; then
```

File: deploy/examples/import-external-cluster.sh
Comment (travisn) [inline deploy/examples/import-external-cluster.sh:41]: How about this?
```suggestion
    echo "For a kernel less than 5.4 use 'layering'."
    echo "For a kernel at least 5.4 use 'layering,fast-diff,image-map,deep-flatten,exclusive-lock'"
```

File: deploy/examples/import-external-cluster.sh
Comment (travisn) [reply deploy/examples/import-external-cluster.sh]: It's amazing how long old kernels linger, but perhaps it is reasonable to assume at least 5.4 as the default. @Madhu-1 wdyt?

File: deploy/examples/import-external-cluster.sh
Comment (anthonyeleven) [reply deploy/examples/import-external-cluster.sh:41]: I'm pushing a new variant

File: deploy/examples/import-external-cluster.sh
Comment (anthonyeleven) [reply deploy/examples/import-external-cluster.sh:39]: 🤐

File: Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md
Comment (anthonyeleven) [reply Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md:62]: Traditional Ceph sets these as a numeric value, so I wanted to make the congruence clear to people used to thinking in those terms.

Comment (anthonyeleven) [reply Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md:62]: Blank lines are kind of inconsistent throughout the project, but I've taken out this one.

File: deploy/charts/rook-ceph-cluster/values.yaml
Comment (anthonyeleven) [reply deploy/charts/rook-ceph-cluster/values.yaml:427]: Done.  Missed that when I reworded.  I obsess a bit over wording.

**Comment by anthonyeleven** [comment]
I believe my latest push addresses the latest comments.

**Comment by anthonyeleven** [comment]
My thoughts were to not implicitly require 5.4+, but also to not be held back by older.  If you've ever run `rbd du` on a 16TB RBD volume you'll understand why.  IMHO defaulting to 5.4+ but calling out what to change for older is a reasonable approach.

File: Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md
Comment (anthonyeleven) [reply Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md]: https://docs.ceph.com/en/latest/rbd/rbd-config-ref/#image-features

It's the sum of each feature's "internal value", which are powers of two.  Old-school bitfields, OR them together, which works out to addition.  Layering alone would be `1`.  This tends to confuse people.

I've amended in what I think is the way you've requested.

Comment (travisn) [reply Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md]: Thanks, makes sense now :)

File: deploy/examples/import-external-cluster.sh
Comment (Madhu-1) [reply deploy/examples/import-external-cluster.sh]: I have seen people still using 3.10 or 4.17 in slack/github issues, assuming 5.4 as default might not be the right one.

File: Documentation/CRDs/Cluster/external-cluster.md
Comment (Madhu-1) [inline Documentation/CRDs/Cluster/external-cluster.md:112]: Do we have an `image-map` image feature? it should be `object-map` isn't it?

File: Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md
Comment (Madhu-1) [inline Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md:67]: Please mention passing this as comma-separated values in imageFeatures parameter.

File: deploy/examples/import-external-cluster.sh
Comment (Rakshith-R) [inline deploy/examples/import-external-cluster.sh:17]:
```suggestion
ROOK_RBD_FEATURES=${ROOK_RBD_FEATURES:-"layering"}
```
I'd prefer we keep it this way, so it is backward compatible and we don't bother
users with having to setting this option explicitly.

File: deploy/examples/import-external-cluster.sh
Comment (Rakshith-R) [inline deploy/examples/import-external-cluster.sh:43]: maybe use this check to for `layering`, since it is a requirement ?

File: deploy/examples/csi/rbd/storageclass-ec.yaml
Comment (Rakshith-R) [inline deploy/examples/csi/rbd/storageclass-ec.yaml:69]:
```suggestion
  imageFeatures: layering,fast-diff,object-map,deep-flatten,exclusive-lock
```

I'd prefer we have only layering by default and list this alongside in a comment, so user can add it as required.
Most users will still be on lower kernel, they will have to edit this when they get an error after creating a PVC. 
This will be very inconvenient according to me.

File: deploy/examples/csi/rbd/storageclass.yaml
Comment (Rakshith-R) [inline deploy/examples/csi/rbd/storageclass.yaml:66]: I'd prefer we have only layering by default and list this alongside in a comment, so user can add it as required.
Most users will still be on lower kernel, they will have to edit this when they get an error after creating a PVC.
This will be very inconvenient according to me.

**Review by Rakshith-R — CHANGES_REQUESTED** [review]

File: deploy/examples/import-external-cluster.sh
Comment (anthonyeleven) [reply deploy/examples/import-external-cluster.sh]: 3.10 is RHEL/CentOS, AIUI that will have various newer code cherry-picked. 

I still think 5.4+ as a baseline is reasonable, given that I've documented the dependency and what to do if one is stuck with an older kernel.  There will be a long tail of stragglers *forever*, at some point we have to move on.

File: Documentation/CRDs/Cluster/external-cluster.md
Comment (anthonyeleven) [reply Documentation/CRDs/Cluster/external-cluster.md:112]: My ignominy is without bound. You're right, thanks.  Doing too many things at once with too many distractions :-/  This is why we have PR review.

File: Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md
Comment (anthonyeleven) [reply Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md:67]: I'm pushing elaboration. PTAL

File: deploy/examples/import-external-cluster.sh
Comment (anthonyeleven) [reply deploy/examples/import-external-cluster.sh:17]: I *almost* did it that way, but had interpreted the request the way I had it.  Pushing this change.

File: deploy/examples/import-external-cluster.sh
Comment (anthonyeleven) [reply deploy/examples/import-external-cluster.sh]: I'm not sure that it is actually a requirement, and we don't sanity-check any other arguments in this section.  Nonetheless I'm pushing an update.

File: deploy/examples/csi/rbd/storageclass-ec.yaml
Comment (anthonyeleven) [reply deploy/examples/csi/rbd/storageclass-ec.yaml:69]: I think today most users will be a kernel that's new enough ; users of older installations will still have the old script.

File: deploy/examples/csi/rbd/storageclass.yaml
Comment (anthonyeleven) [reply deploy/examples/csi/rbd/storageclass.yaml:66]: Hrm, my reply seems to have vanished.  I don't think "most" users will be on a pre-5.4 kernel. At some point we have to draw the line, why not now?  It's easy enough for someone to edit, though features can be added per-volume after the fact too via the Ceph CLI.  I'm changing it to as you request; we'll still be improving the outdated info currently in the various files.

**Comment by anthonyeleven** [comment]
The failed unit doesn't _look_ like it is related to this change

File: deploy/examples/csi/rbd/storageclass-ec.yaml
Comment (travisn) [reply deploy/examples/csi/rbd/storageclass-ec.yaml:69]: Looks like the other storage classes default to `layering`, and just the EC storage class defaults to the newer kernel features? Seems like we should be consistent with the defaults. 
To @Rakshith-R's point, the consequence of trying to run the defaults is painful since PVs will fail if you have an older kernel and don't know to change this. With the `layering` default, everything still works even if not optimally. We don't know how many users are on older vs newer kernels, but I'd feel ok going with the newer default if we felt like 95% of users are on the newer kernel. It still feels like we are not there. 
What if we add a note in the [Prerequisites doc](https://rook.github.io/docs/rook/latest/Getting-Started/Prerequisites/prerequisites/#kernel) that they should update the storage class if they have a 5.4+ kernel?

Comment (anthonyeleven) [reply deploy/examples/csi/rbd/storageclass-ec.yaml]: Fair enough.  I've tweaked accordingly.  PTAL.

File: deploy/examples/csi/rbd/storageclass.yaml
Comment (anthonyeleven) [reply deploy/examples/csi/rbd/storageclass.yaml]: I've pushed changes, PTAL

**Comment by anthonyeleven** [comment]
@Rakshith-R @travisn the failed tests appear to be unrelated to these changes

**Review by travisn — APPROVED** [review]

File: deploy/charts/rook-ceph-cluster/values.yaml
Comment (Madhu-1) [inline deploy/charts/rook-ceph-cluster/values.yaml:426]:
```suggestion
        # support only the `layering` feature. The Linux kernel (KRBD) supports the
```

File: deploy/examples/csi/rbd/storageclass-ec.yaml
Comment (Madhu-1) [inline deploy/examples/csi/rbd/storageclass-ec.yaml:67]:
```suggestion
  # support only the `layering` feature. The Linux kernel (KRBD) supports the
```

File: deploy/examples/csi/rbd/storageclass.yaml
Comment (Madhu-1) [inline deploy/examples/csi/rbd/storageclass.yaml:64]:
```suggestion
  # support only the `layering` feature. The Linux kernel (KRBD) supports the
```

File: Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md
Comment (Madhu-1) [inline Documentation/Storage-Configuration/Block-Storage-RBD/block-storage.md:64]:
```suggestion
    # support only the `layering` feature. The Linux kernel (KRBD) supports the
```

File: Documentation/Getting-Started/Prerequisites/prerequisites.md
Comment (Madhu-1) [inline Documentation/Getting-Started/Prerequisites/prerequisites.md:102]:
```suggestion
Rook's default RBD configuration specifies only the `layering` feature, for
broad compatibility with older kernels. If your Kubernetes nodes run a 5.4
or later kernel you may wish to enable additional feature flags. The `fast-diff`
and `object-map` features are especially useful.
```

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
nits

File: Documentation/Getting-Started/Prerequisites/prerequisites.md
Comment (anthonyeleven) [reply Documentation/Getting-Started/Prerequisites/prerequisites.md]: I've amended with double spaces collapsed.

**Comment by Madhu-1** [comment]
@anthonyeleven can you please rebase the PR? changes LGTM

**Comment by travisn** [comment]
@anthonyeleven The changes look ready for merge after the merge conflict is resolved. Still working on it?

**Comment by anthonyeleven** [comment]
@travisn Still dealing with cross-country move fallout.

Thought I had to rebase, which I admittedly have never managed successfully - I always end up dragging in a thousand unrelated commits.

But it looks like it was a single word conflict, I’ve resolved that via the GUI.

There are two changes requested from @Madhu-1  and @Rakshith-R  that I think I’ve addressed.  If they aren’t fully, I’d love to just merge this finally and I can address in a followup.

**Review by Madhu-1 — APPROVED** [review]

**Comment by Madhu-1** [comment]
@travisn good to merge this one?

File: deploy/examples/import-external-cluster.sh
Comment (parth-gr) [inline deploy/examples/import-external-cluster.sh:17]: If this can be changed, we can have a flag from python script(external script) to update this value?

**Comment by travisn** [comment]
> @travisn good to merge this one?

Yes, will dismiss Rakshith's review since it's addressed

**Pull request merged**
